src/embassy/time_driver.rs

```diff
@@ -0,0 +1,394 @@
+// This allows for now() to return the correct value even if it races an overflow.
+//
+// To get `now()`, `period` is read first, then `counter` is read. If the counter value matches
+// the expected range for the `period` parity, we're done. If it doesn't, this means that
+// a new period start has raced us between reading `period` and `counter`, so we assume the `counter` value
+// corresponds to the next period.
+//
+// `period` is a 32bit unsigned integer, so It overflows on 2^32 * 2^15 / 32768 seconds of uptime, which is 136 years.
+fn calc_now(period: u32, counter: u32) -> u64 {
+    ((period as u64) << 15) + ((counter ^ ((period & 1) << 15)) as u64)
+}
+
+const ALARM_STATE_NEW: AlarmState = AlarmState::new();
+
+pub(super) struct EmbassyTimeDriver {
+    /// Number of 2^31 periods elapsed since boot.
+    period: AtomicU32,
+    alarm_count: AtomicU8,
+    /// Timestamp at which to fire alarm. u64::MAX if no alarm is scheduled.
+    alarms: Mutex<CriticalSectionRawMutex, [AlarmState; ALARM_COUNT]>,
+}
+
+impl EmbassyTimeDriver {
+    pub(super) fn init(timer: Timer, clocks: &Clocks, apb: &mut Bus) {
+        Timer::enable(apb);
+        Timer::reset(apb);
+
+        timer.ctl0().modify(|_, w| w.cen().disabled());
+
+        timer.cnt().write(|w| w.cnt().bits(0));
+
+        let timer_freq = <Timer as RcuBus>::Bus::get_timer_frequency(clocks);
+        // NOTE: using the default 1MHz tick rate
+        let psc = (timer_freq.0 as u64) / TICK_HZ - 1;
+        let psc: u16 = match psc.try_into() {
+            Err(_) => panic!("psc division overflow: {}", psc),
+            Ok(n) => n,
+        };
+
+        timer.psc().write(|w| w.psc().bits(psc));
+        timer.car().write(|w| {
+            w.car()
+                .bits(u16::MAX as <RegisterBlock as ChannelsAccess>::Counter)
+        });
+
+        timer.ctl0().modify(|_, w| w.ups().counter_only());
+        timer.swevg().write(|w| w.upg().update());
+        timer.ctl0().modify(|_, w| w.ups().any_event());
+
+        // Half of the u16::MAX value
+        timer.ch0cv().write(|w| w.ch0val().bits(0x8000));
+
+        timer
+            .dmainten()
+            .modify(|_, w| w.upie().enabled().ch0ie().enabled());
+
+        #[cfg(feature = "time-driver-tim1")]
+        {
+            unsafe {
+                cortex_m::peripheral::NVIC::unmask(Interrupt::TIMER1);
+            }
+        }
+        #[cfg(feature = "time-driver-tim2")]
+        {
+            unsafe {
+                cortex_m::peripheral::NVIC::unmask(Interrupt::TIMER2);
+            }
+        }
+        #[cfg(feature = "time-driver-tim14")]
+        {
+            unsafe {
+                cortex_m::peripheral::NVIC::unmask(Interrupt::TIMER14);
+            }
+        }
+
+        timer.ctl0().modify(|_, w| w.cen().enabled());
+    }
+
+    fn get_alarm<'a>(&'a self, cs: CriticalSection<'a>, alarm: AlarmHandle) -> &'a AlarmState {
+        // safety: we're allowed to assume the AlarmState is created by us, and
```

Comment:
Do you mean the `AlarmHandle`? This is a safe method, so you can't assume anything about what arguments it is called with. Either make this method `unsafe` and document the safety requirement, or use a safe method below to panic if the handle is out of bounds rather than causing undefined behaviour.